### PR TITLE
feat: notify on successful autoupdate

### PR DESCRIPTION
**Pull request by mtojek**
Related: https://github.com/coder/coder/pull/13891

This PR implements notifications on workspace autoupdate.

Comment:
I took inspiration from [Bruno's PR](https://github.com/coder/coder/pull/13868), and exposed the `FakeNotificationsEnqueuer` as a separate package.

### CODE REVIEW

File: coderd/autobuild/lifecycle_executor.go
Comment (BrunoQuaresma) [inline coderd/autobuild/lifecycle_executor.go:301]: **Learning:** Why do we use `log.Warn` instead of `slog.Warn` and only use `slog` to print the error?

File: coderd/coderdtest/coderdtest.go
Comment (BrunoQuaresma) [inline coderd/coderdtest/coderdtest.go:246]: Learning: Is it the same of `&notiffake.FakeNotificationEnqueuer{}`?

**Review by BrunoQuaresma — COMMENTED** [review]
Overall, the code looks good to me, but I would wait for @dannykopping review since he is leading the work on this feature.

File: coderd/autobuild/lifecycle_executor.go
Comment (mtojek) [reply coderd/autobuild/lifecycle_executor.go:301]: If you look at the imports, it is actually slog package: `cdr.dev/slog`. I'm not sure if I have seen `slog.Warn` in the project? Here, `log` is just an instance of slog logger :)

File: coderd/autobuild/lifecycle_executor.go
Comment (dannykopping) [inline coderd/autobuild/lifecycle_executor.go:297]:
```suggestion
						}, "autobuild",
```

File: coderd/autobuild/lifecycle_executor.go
Comment (dannykopping) [inline coderd/autobuild/lifecycle_executor.go:217]: I think this flag could be more clearly named; as it is right now it has no real conceptual link to notifications, so perhaps something like `didAutoUpdate` or something would be more clear? We could also add a comment here to explain why we're capturing this state (so that we can know when a workspace was auto-updated and notify accordingly).

File: coderd/database/migrations/000227_notifications_workspace_autoupdated.up.sql
Comment (dannykopping) [inline coderd/database/migrations/000227_notifications_workspace_autoupdated.up.sql:3]: Perhaps we should explain that this would have restarted it, as well?

File: coderd/notifications/notiffake/notiffake.go
Comment (dannykopping) [inline coderd/notifications/notiffake/notiffake.go:1]: Nit: I like splitting this out but I think the name is too "cute".
I think something like `testutil` or something would be more idiomatic.

**Review by dannykopping — APPROVED** [review]
LGTM!

File: coderd/coderdtest/coderdtest.go
Comment (mtojek) [reply coderd/coderdtest/coderdtest.go:246]: Yes, it is just a different form of writing it 👍

File: coderd/notifications/notiffake/notiffake.go
Comment (mtojek) [reply coderd/notifications/notiffake/notiffake.go:1]: Yes, I'm aware of it. This is an alternative approach, and I'm happy to leave it or switch to the other form. @dannykopping any preference?

Comment (dannykopping) [reply coderd/notifications/notiffake/notiffake.go:1]: I prefer names which are explicit and clear, so I prefer `testutils` TBH.

Comment (mtojek) [reply coderd/notifications/notiffake/notiffake.go:1]: I will wait with merging this PR until Bruno merges https://github.com/coder/coder/pull/13868

File: coderd/autobuild/lifecycle_executor.go
Comment (mtojek) [reply coderd/autobuild/lifecycle_executor.go]: Fixed, switched to `didAutoUpdate` with a comment

File: coderd/autobuild/lifecycle_executor.go
Comment (mtojek) [reply coderd/autobuild/lifecycle_executor.go]: Fixed, nice catch

File: coderd/database/migrations/000227_notifications_workspace_autoupdated.up.sql
Comment (mtojek) [reply coderd/database/migrations/000227_notifications_workspace_autoupdated.up.sql:3]: Oh, I think it wasn't restarted due to the update, but rather updated on the autostart. There is a configuration option to force the latest template version.

Comment (dannykopping) [reply coderd/database/migrations/000227_notifications_workspace_autoupdated.up.sql:3]: How can we update a template without restarting a workspace?

Comment (mtojek) [reply coderd/database/migrations/000227_notifications_workspace_autoupdated.up.sql:3]: My assumption is as follows:

when the workspace is stopped by user or auto-stopped, and the next morning Lifecycle Executor brings it back to life, it will force the latest template version. I don't think we have a feature that stops the workspace and restarts with the latest template version?

Comment (dannykopping) [reply coderd/database/migrations/000227_notifications_workspace_autoupdated.up.sql:3]: Oh I see what you mean! Gotcha.

File: coderd/notifications/notiffake/notiffake.go
Comment (mtojek) [reply coderd/notifications/notiffake/notiffake.go:1]: Alright, I decided to merge it as is, and refactor/unify once https://github.com/coder/coder/pull/13868 is merged.

**Pull request merged**
